Analysis/Core/include/AnalysisCore/TrackSelectorPID.h

```diff
@@ -252,9 +252,159 @@ class TrackSelectorPID
+  template <typename T>
+  bool isSelectedTrackPIDRICH(const T& track, bool& conditionalTOF)
+  {
+    // Accept if selection is disabled via large values.
+    if (mNSigmaRICHMin < -999. && mNSigmaRICHMax > 999.) {
+      return true;
+    }
+
+    // Get nσ for a given particle hypothesis.
+    double nSigma = 100.;
+    switch (mPdg) {
+      case kElectron: {
+        nSigma = track.rich().richNsigmaEl();
+        break;
+      }
+      case kMuonMinus: {
+        nSigma = track.rich().richNsigmaMu();
+        break;
+      }
+      case kPiPlus: {
+        nSigma = track.rich().richNsigmaPi();
+        break;
+      }
+      case kKPlus: {
+        nSigma = track.rich().richNsigmaKa();
+        break;
+      }
+      case kProton: {
+        nSigma = track.rich().richNsigmaPr();
+        break;
+      }
+      default: {
+        LOGF(error, "ERROR: RICH PID not implemented for PDG %d", mPdg);
+        assert(false);
+      }
+    }
+
+    if (mNSigmaRICHMinCondTOF < -999. && mNSigmaRICHMaxCondTOF > 999.) {
+      conditionalTOF = true;
+    } else {
+      conditionalTOF = mNSigmaRICHMinCondTOF <= nSigma && nSigma <= mNSigmaRICHMaxCondTOF;
+    }
+    return mNSigmaRICHMin <= nSigma && nSigma <= mNSigmaRICHMax;
+  }
+
+  /// Returns status of RICH PID selection for a given track.
+  /// \param track  track
+  /// \return RICH selection status (see TrackSelectorPID::Status)
+  template <typename T>
+  int getStatusTrackPIDRICH(const T& track)
+  {
+    if (isValidTrackPIDRICH(track)) {
+      bool condTOF = false;
+      if (isSelectedTrackPIDRICH(track, condTOF)) {
+        return Status::PIDAccepted; // accepted
+      } else if (condTOF) {
+        return Status::PIDConditional; // potential to be accepted if combined with TOF
+      } else {
+        return Status::PIDRejected; // rejected
+      }
+    } else {
+      return Status::PIDNotApplicable; // PID not applicable
+    }
+  }
+
+  // MID
+
+  /// Checks if track is OK for MID PID.
+  /// \param track  track
+  /// \return true if track is OK for MID PID
+  template <typename T>
+  bool isValidTrackPIDMID(const T& track)
+  {
+    return track.midId() > -1;
+  }
+
+  /// Checks if track is compatible with muon hypothesis in the MID detector.
+  /// \param track  track
+  /// \return true if track has been identified as muon by the MID detector
+  template <typename T>
+  bool isSelectedTrackPIDMID(const T& track)
+  {
+    if (mPdg != kMuonMinus) {
+      return false;
+    }
+    return track.mid().midIsMuon() == 1; // FIXME: change to return track.midIsMuon() once the column is bool.
+  }
+
+  /// Returns status of MID PID selection for a given track.
+  /// \param track  track
+  /// \return MID selection status (see TrackSelectorPID::Status)
+  template <typename T>
+  int getStatusTrackPIDMID(const T& track)
+  {
+    if (mPdg != kMuonMinus) {
+      return Status::PIDRejected;
+    }
+    if (isValidTrackPIDMID(track)) {
```

Comment:
As a general comment I would try to code things so that you bail out as soon as possible. In this particular case:

```
if (!isValidTrackPIDMID(track)) {
  return Status::PIDNotApplicable; // PID not applicable
}
```

Reply:
hi @ktf, from what Vit told me, there is still an issue when trying to run the Jpsi/X analysis with Run2/3 data that is investigating that could be even a more general problem. Therefore, he also agrees to go ahead and merge it. Thanks.

Reply:
Agreed, thanks.